### DISCUSSION TITLE
Runtime size vs. type-derived size

### DISCUSSION THREAD

**Discussion by lthls** [ocaml/ocaml#11895] [General]
(Note: this issue is meant as a discussion, not a bug report. I plan to close it just after submitting it. I've thought about using other means of communication, but Github looks like the place most likely to be looked at by people interested in the discussion.)

We had a few discussions with colleagues about whether the compiler should support runtime values that are bigger than their type suggests. Currently it can only be done with `Obj.magic`, except for a corner case described at the end of this post.

Here is a typical use case:
```ocaml
type point2d = { x : float; y : float; }
type point3d = { x : float; y : float; z : float; }

let project2d_unsafe (p3 : point3d) : point2d = Obj.magic p3

let project2d_array_unsafe (v3 : point3d array) : point2d array = Obj.magic v3
```
It's obviously not something we want to encourage, in particular because the extra fields could become unreachable from the code but the GC will not be able to collect them. But I would be surprised if there's not at least one piece of code using a trick like that in the wild.

Currently, if we except the issues with collecting unreachable values (and potential extra memory usage) there doesn't seem to be anything in the compiler relying on the fact that runtime values are not bigger than their type would suggest, so the above code should work as expected. But there is (almost, see end of the post for details) nothing that allows one to produce a value whose runtime size is bigger than the size for its type without the `Obj` module either, so we're still in a position where we can forbid it.

So we were left wondering: should we explicitly support the feature ? Explicitly forbid it ? Do nothing ?
The flambda 2 code is moving towards the second solution; in theory it means that we could treat code that is currently legal (for example, `let f (x, y) = ... in f (Obj.magic (0,1,2))`) as if it was unreachable. In practice this interpretation is most useful in cases like `let option_of_error = function | Ok x -> Some x | Error _ -> None`, where the strict semantics allow us to replace the allocation of `Some x` by the input value (see also #8958 for a version based on flambda 1). In a less strict semantics this could silently keep more values alive, so would not be obviously beneficial or even correct.
However, supporting mismatching sizes could be useful to implement new features/optimisations. For example, an explicit no-cost coercion operator for modules (`Set.Make(MyModule :> Set.OrderedType)`). There were also a few similar opportunities that arose during a past internship about identity functions.

To conclude, here is an example that illustrates the only case I'm aware of where the compiler can produce values with mismatched sizes for genuine code:
```ocaml
(* m.mli *)
val x : int

(* m.ml *)
let x = 0
let y = 1

(* test.ml *)
module type S = sig val x : int end

let m1 = (module M : S)
let m2 = (module struct let x = 0 let y = 1 end : S)
let m3 = (module struct let x = 0 end : S)

let _ =
  Format.printf "m1 = m2 => %b@.m1 = m3 => %b@.m2 = m3 => %b@."
    (m1 = m2)
    (m1 = m3)
    (m2 = m3)
```
The example prints `true` in all cases in bytecode and with flambda, but prints `false` in the first two cases with closure. This is because the special compilation scheme for compilation unit module blocks in Closure may append extra fields at the end of the module block, after the ones in the signature (in this case the value of `y`).

**Comment by trefis** [comment]
> But I would be surprised if there's not at least one piece of code using a trick like that in the wild.

For the record, merlin used to do [this](https://github.com/ocaml/merlin/blob/bad59c1dfd9a149ff4210bb2441144abce873d34/src/ocaml_aux/parsing_aux.ml#L26); but we've moved away from that many years ago.

**Comment by gasche** [comment]
For me the behavior reported with first-class module is a bug that should be fixed: `(module M : S)` should have the size of `S`, not the size of `M`, and I would expect a shallow copy if they mismatch.

Two questions:
- Is there a good reason to do it this way (this turns out to be important for performance in some common cases I am not thinking of), or this is merely an accident?
- How hard would it be to fix this?

**Reply by lthls** [reply]
> * Is there a good reason to do it this way (this turns out to be important for performance in some common cases I am not thinking of), or this is merely an accident?

I don't know the reason.

> * How hard would it be to fix this?

It's very easy to fix it in a way that makes performance worse (just use the normal translation scheme, like Flambda).
@chambart thinks that it should not be too hard to fix it in a performance-preserving way, and I tend to agree with him, but we haven't actually tried it yet.
